packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+            _isBuy
+        );
+    }
+
+    /**
+     * @notice Extracts expired locks and deletes them.
+     * @param _subject Subject address for which locks are being extracted.
+     * @param _indexes Indexes of the locks to be extracted.
+     * @return subjectToken_ Address of the subject token.
+     * @return totalAmount_ Total amount of tokens withdrawn.
+     */
+    function _extractExpiredAndDeleteLocks(
+        address _subject,
+        uint256[] memory _indexes
+    ) internal returns (address subjectToken_, uint256 totalAmount_) {
+        if (_isZeroAddress(_subject)) {
+            revert Staking_InvalidSubject();
+        }
+
+        if (_indexes.length == 0) {
+            revert Staking_EmptyIndexes();
+        }
+        LockInfo memory lockInfo = locks[_indexes[0]];
+        subjectToken_ = lockInfo.subjectToken;
+
+        for (uint256 i = 0; i < _indexes.length; i++) {
+            uint256 index = _indexes[i];
+            lockInfo = locks[index];
+            if (lockInfo.subject != _subject) {
+                revert Staking_SubjectsDoesNotMatch(index);
+            }
+            if (lockInfo.unlockTimeInSec > block.timestamp) {
+                revert Staking_LockNotExpired(
+                    index,
+                    block.timestamp,
+                    lockInfo.unlockTimeInSec
+                );
+            }
+            if (lockInfo.user != msg.sender) {
+                revert Staking_NotOwner(index);
+            }
+            totalAmount_ += lockInfo.amount;
+            delete locks[index];
+        }
+    }
+
+    /**
+     * @notice Deposits and locks tokens for a single subject.
+     * @param _subject Subject address for which tokens are being deposited.
+     * @param _amount Amount of tokens being deposited.
+     * @param _lockPeriodInSec Lock period for the tokens.
+     */
+    function _depositAndLock(
+        address _subject,
+        uint256 _amount,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    ) internal returns (uint256 unlockTimeInSec_) {
+        IERC20Extended subjectToken;
+        (subjectToken, unlockTimeInSec_) = _createLock(
+            _subject,
+            _amount,
+            _lockPeriodInSec,
+            _beneficiary,
+            false
+        );
+
+        // Transfer the tokens to this contract
+        subjectToken.safeTransferFrom(msg.sender, address(this), _amount);
+    }
+
+    /**
+     * @notice Sets the lock period for staking. only owner can call this function.
+     * @param _lockPeriodInSec Lock period to set.
+     * @param _allowed Boolean to allow or disallow the lock period.
+     */
+    function setLockPeriod(
+        uint256 _lockPeriodInSec,
+        bool _allowed
+    ) external onlyRole(CHANGE_LOCK_DURATION) {
+        if (lockPeriodsInSec[_lockPeriodInSec] == _allowed) {
+            revert Staking_LockPeriodAlreadySet();
+        }
+        lockPeriodsInSec[_lockPeriodInSec] = _allowed;
```

Comment:
may be we can rename this lockPeriodsInSec to represent map. similar name is used for variable and map